### PR TITLE
Move the histogram class into its own files.

### PR DESCRIPTION
**Pull request by pgavlin**
As it says. This is in preparation for bringing this code up-to-date
with the JIT CC.

### CODE REVIEW

**Comment by pgavlin** [comment]
@dotnet/jit-contrib PTAL

**Comment by adityamandaleeka** [comment]
Please add license headers to the new files.

**Comment by pgavlin** [comment]
@adityamandaleeka: good catch. Done.

**Comment by adityamandaleeka** [comment]
Thanks for adding the headers. I'm not familiar with this code. Is this check:

`#if CALL_ARG_STATS || COUNT_BASIC_BLOCKS || COUNT_LOOPS || EMITTER_STATS || MEASURE_NODE_SIZE` 

no longer needed?

Other than that, LGTM if CI passes.

**Comment by pgavlin** [comment]
Those checks should not be required at the site of the type definition. They are still relevant (and present) at its uses.

Unfortunately, this change may need to wait: MSVC is failing to compile with an internal compiler error.

**Comment by adityamandaleeka** [comment]
> Those checks should not be required at the site of the type definition. They are still relevant (and present) at its uses.

Sounds good. I assumed that was the case, but wanted to double-check that it was intentional.

File: src/jit/histogram.hpp [hunk 1]
```diff
@@ -0,0 +1,29 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#ifndef _HISTOGRAM_HPP_
```
Comment (BruceForstall) [inline src/jit/histogram.hpp]: This file should be named histogram.h, not .hpp.

**Comment by BruceForstall** [comment]
> Those checks should not be required at the site of the type definition. They are still relevant (and present) at its uses.

The reason it is useful to have them at the type definition is to avoid accidentally including the class in any build that doesn't expect to have it. I would suggest adding them back, as documentation of that goal.

**Comment by pgavlin** [comment]
> The reason it is useful to have them at the type definition is to avoid accidentally including the class in any build that doesn't expect to have it.

Is there a measurable impact on the built JIT if this type is included?

**Comment by BruceForstall** [comment]
If it's unused, the linker should rip it out. The philosophy is simply to be intentional when putting something in the release build.

**Comment by pgavlin** [comment]
AFAICT there is no reasonable workaround for the C++ FE bug I'm seeing, so I'm going to close this for now.

**Pull request closed without merging**